clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -156,8 +156,9 @@ public void authenticate() throws IOException {
                 // API_VERSIONS_REQUEST must be sent prior to sending SASL_HANDSHAKE_REQUEST to
                 // fetch supported versions.
                 String clientId = (String) configs.get(CommonClientConfigs.CLIENT_ID_CONFIG);
-                currentRequestHeader = new RequestHeader(ApiKeys.SASL_HANDSHAKE.id, clientId, correlationId++);
                 SaslHandshakeRequest handshakeRequest = new SaslHandshakeRequest(mechanism);
+                currentRequestHeader = new RequestHeader(ApiKeys.SASL_HANDSHAKE.id,
+                        handshakeRequest.getVersion(), clientId, correlationId++);
```

Comment:
Using the current version is ok for now since there is currently only one version of SaslHandshakeRequest. It would be useful to add a comment that when there are new versions of SaslHandshakeRequest, we need to select the version based on the result of ApiRequest accordingly.